Cython/Compiler/ExprNodes.py

```diff
@@ -6628,20 +6628,56 @@ class PyMethodCallNode(CallNode):
     # Specialised call to a (potential) PyMethodObject with non-constant argument tuple.
     # Allows the self argument to be injected directly instead of repacking a tuple for it.
     #
-    # function    ExprNode      the function/method object to call
+    # function    ExprNode  the function/method object to call
     # arg_tuple   TupleNode     the arguments for the args tuple
     # kwdict      ExprNode or None  keyword dictionary (if present)
     # unpack      bool
+    # function_obj ExprNode or None   
 
     subexprs = ['function', 'arg_tuple', 'kwdict']
     is_temp = True
     kwdict = None
+    avoid_attribute_lookup = False
+    function_obj = None
+
+    def __init__(self, pos, **kwds):
+        super().__init__(pos, **kwds)
+        if self.can_avoid_attribute_lookup():
+            self.avoid_attribute_lookup = True
+            self.function_obj = self.function.obj
+
+    def can_avoid_attribute_lookup(self):
+        # Essentially, if the signature matches PyObject_VectorcallMethod
+        # then it's worth doing that directly and not creating a new method in
+        # the attribute lookup.
+        if self.kwdict and not isinstance(self.kwdict, DictNode):
+            return False
+        if not self.function.is_attribute:
+            return False
+        # These two determine that it's not just a plain getattr
+        if not self.function.is_py_attr:
+            return False
+        if self.function.is_special_lookup:
+            return False
+        if not self.attribute_is_likely_method(self.function):
+            # PyObject_VectorcallMethod would work, but is more likely to
+            # be a pessimization.
+            return False
+        return True
+
+    @staticmethod
+    def attribute_is_likely_method(attr):
+        obj = attr.obj
+        if obj.is_name and obj.entry.is_pyglobal:
+            return False  # more likely to be a function
+        return True
 
     def generate_evaluation_code(self, code):
```

Comment:
This method is a lengthy mess of conditional code that could really benefit from getting split up into several smaller methods. That would allow two different implementations without all the new "if avoid attribute lookup" special casing, which just adds to the existing complexity. Actually, more than two implementations, "having kwargs as key-value pairs" is another very different case that merits its own separate method.

I can't really say if this PR is a good place to do this split, but it feels very worthwhile.

Reply:
I've had a go at splitting it up. It's still all a bit tightly coupled, but it's hopefully a bit more readable.

Reply:
I think I'd even try to go a step further and use two different classes for the two cases of `PyVectorcall_Method()` and regular method lookup/unpacking/calling. Maybe with a factory method to select (and instantiate) the right node class for the given configuration. Maybe with a common base class if needed to share functionality, although that seems limited overall. Simple parts might just end up as functions.